delfin/drivers/dell_emc/vmax/client.py

```diff
@@ -427,6 +427,202 @@ def list_disks(self, storage_id):
+            for host in hosts:
+                host_info = self.rest.get_host(
+                    self.array_id, self.uni_version, host)
+
+                host_item = {
+                    'storage_id': storage_id,
+                    'native_storage_host_id': host_info.get('hostId'),
+                    'name': host_info.get('hostId'),
+                    'status': constants.HostStatus.NORMAL,
+                }
+                host_list.append(host_item)
+            return host_list
+
+        except Exception:
+            LOG.error("Failed to get storage host details from VMAX")
+            raise
+
+    def list_storage_host_groups(self, storage_id):
+        try:
+            # Get list of storage host groups
+            host_groups = self.rest.get_host_group_list(self.array_id,
+                                                        self.uni_version)
+            host_group_list = []
+            storage_host_grp_relation_list = []
+            for host_group in host_groups:
+                host_group_info = self.rest.get_host_group(
+                    self.array_id, self.uni_version, host_group)
+                host_group_item = {
+                    'storage_id': storage_id,
+                    'native_storage_host_group_id': host_group,
+                }
+                host_group_list.append(host_group_item)
+
+                for storage_host in host_group_info['host']:
+                    storage_host_group_relation = {
+                        'storage_id': storage_id,
+                        'native_storage_host_group_id': host_group,
+                        'native_storage_host_id': storage_host.get('hostId')
+                    }
+                    storage_host_grp_relation_list \
+                        .append(storage_host_group_relation)
+
+            result = {
+                'storage_host_groups': host_group_list,
+                'storage_host_grp_host_rels': storage_host_grp_relation_list
+            }
+
+            return result
+
+        except Exception:
+            LOG.error("Failed to get storage host group details from VMAX")
+            raise
+
+    def list_port_groups(self, storage_id):
+        try:
+            # Get list of port groups
+            port_groups = self.rest.get_port_group_list(self.array_id,
+                                                        self.uni_version)
+            port_group_list = []
+            port_group_relation_list = []
+            for port_group in port_groups:
+                port_group_info = self.rest.get_port_group(
+                    self.array_id, self.uni_version, port_group)
+                port_group_item = {
+                    'storage_id': storage_id,
+                    'native_port_group_id': port_group,
+                }
+                port_group_list.append(port_group_item)
+
+                for port in port_group_info['symmetrixPortKey']:
+                    port_name = port['directorId'] + ':' + port['portId']
```

Comment:
Just portId is not unique, can it be not treated as native_port_id?

Reply:
> Just portId is not unique, can it be not treated as native_port_id?

Yes, portId is not unique, so we generate directorId + portId for native_port_id

Reply:
OK